README.rst
=====
ionex
=====

Модуль для чтения карт в формате IONEX.


*************
Использование
*************

Коротко
-------

::

    import ionex

    with open('igsg0010.00i') as file:
        inx = ionex.reader(file)
        for ionex_map in inx:
            print(ionex_map.epoch)
            print(ionex_map.tec)


Содержание модуля
------------------


~~~~~~~~~~~~~~~~~~~~
`ionex.reader(file)`
~~~~~~~~~~~~~~~~~~~~

Возвращает читалку файла в формате IONEX. Читалка - итерируемый объект, который
на каждой итерации возвращает экземпляр `IonexMap` очередной карты, прочитанной
из файла.

**Параметры**

- `file`: `str` | `file`, путь к файлу IONEX или объект файла.

**Исключения**

- `IONEXError`, неизвестный тип или версия переданного файла.
- `IONEXUnexpectedEnd`, неполный файл.
- `IONEXMapError`, ошибки при обработке карты.


~~~~~~~~~~~~~~~~~~~~~~~
`class ionex.IonexMap`
~~~~~~~~~~~~~~~~~~~~~~~

Класс карты IONEX, содержит карту значений ПЭС и мета-данные.

**Атрибуты**


- `grid`: `namedtuple`, определение сетки для карты, содержит два `namedtupla`:

  - `grid.latittude` = `namedtuple('Latitude', ['lat1', 'lat2', 'dlat'])`
    определение сетки по широте от `lat1` до `lat2` с шагом `dlat`;

  - `grid.longitude` = `namedtuple('Longitude', ['lon1', 'lon2', 'dlon'])`
    определение сетки по долготе от `lon1` до `lon2` с шагом `dlon`.

- `tec`: `list`, данные ПЭС; одномерный список, представляет собой
  набор широтных "срезов" со значениями ПЭС.

  Начало каждого среза соответствует широте `grid.latitude.lat1`, конец --
  `grid.latitude.lat2`, с шагом, равным `grid.latitude.dlat`.

  Долгота первого среза соответствует `grid.longitude.lon1`, долгота
  последнего -- `grid.longitude.lon2`, с шагом, равным `grid.longitude.dlon`.

- `height`: `float`, высота, с которой ассоциированы данные карты.

- `epoch`: `datetime`, дата и время карты ПЭС.

*********
Установка
*********

Сейчас лучше устанавливать в editable-mode::

    $ pip install -e git+https://github.com/gnss-lab/ionex.git#egg=ionex


******
Ошибки
******

Ошибки стоит оформлять на `issues <https://github.com/gnss-lab/ionex/issues>`_.

Как оформить ошибку:

- опишите, как её можно воспроизвести;
- приложите файлы IONEX, при обработке которых возникла
  ошибка или укажите ссылку на эти файлы.

********
Лицензия
********

Distributed under the terms of the
`MIT <https://github.com/gnss-lab/gnss-tec/blob/master/LICENSE.txt>`_
license, gnss-tec is free and open source software.

Copyright Ilya Zhivetiev, 2018.


CHANGELOG.rst
ionex v0.2
==========

Bug fixes
---------

- Исправлено: возникает ошибка, когда значение часа в строке с эпохой равно 24.


tests/conftest.py
import os
from io import StringIO
from contextlib import contextmanager

import pytest

TEST_DATA_DIR = os.path.join(
    os.path.dirname(os.path.realpath(__file__)),
    'test_data',
)


@contextmanager
def get_file_object(filename):
    file_object = None
    try:
        file_path = os.path.join(TEST_DATA_DIR, filename)
        file_object = open(file_path)
        yield file_object
    finally:
        if file_object is not None:
            file_object.close()


@pytest.fixture
def ionex_file_object():
    with get_file_object('ionex_file.00i') as file_object:
        yield file_object


@pytest.fixture(params=['filename', 'fileobject'])
def ionex_file(request):
    if request.param == 'filename':
        return os.path.join(TEST_DATA_DIR, 'ionex_file.00i')

    if request.param == 'fileobject':
        return request.getfixturevalue('ionex_file_object')


@pytest.fixture
def ionex_file_no_end_of_file():
    with get_file_object('ionex_file.00i') as file_object:
        content = file_object.readlines()
    content.pop(-1)
    file_object = StringIO(''.join(content))
    return file_object


@pytest.fixture
def one_map_file():
    file_obj = None
    try:
        file_path = os.path.join(TEST_DATA_DIR, 'one_map.xxi')
        file_obj = open(file_path)
        yield file_obj
    finally:
        if file_obj is not None:
            file_obj.close()


@pytest.fixture
def one_map_file_data():
    file_path = os.path.join(TEST_DATA_DIR, 'one_map_data.txt')
    with open(file_path) as file_obj:
        contents = file_obj.read()
    data = eval(contents)
    return data


tests/ionex_test.py
from datetime import datetime
from io import StringIO

import pytest

from ionex.ionex_file import IonexV1, Grid, MapGridDef
from ionex.ionex_file import Latitude, Longitude, Height, Map


@pytest.fixture
def ionex_header():
    return StringIO('''\
     1.0            IONOSPHERE MAPS     MIX                 IONEX VERSION / TYPE
tecrms2ionex_2.awk  gAGE/UPC            11/27/07  811UT     PGM / RUN BY / DATE
  1999     1     2     1     0     0                        EPOCH OF FIRST MAP
  1999     1     2    23     0     0                        EPOCH OF LAST MAP
  7200                                                      INTERVAL
    12                                                      # OF MAPS IN FILE
  COSZ                                                      MAPPING FUNCTION
     0.0                                                    ELEVATION CUTOFF
Combined TEC calculated as weighted mean of input TEC valuesOBSERVABLES USED
     0                                                      # OF STATIONS
    27                                                      # OF SATELLITES
  6371.0                                                    BASE RADIUS
     2                                                      MAP DIMENSION
   450.0 450.0   0.0                                        HGT1 / HGT2 / DHGT
    87.5 -87.5  -2.5                                        LAT1 / LAT2 / DLAT
  -180.0 180.0   5.0                                        LON1 / LON2 / DLON
    -1                                                      EXPONENT
                                                            END OF HEADER
     1                                                      START OF TEC MAP
  1999     1     2     1     0     0                        EPOCH OF CURRENT MAP
''')


@pytest.fixture
def lat_def():
    return Latitude(87.5, -87.5, -2.5)


@pytest.fixture
def lon_def():
    return Longitude(-180.0, 180.0, 5.0)


@pytest.fixture
def hgt_def():
    return Height(450.0, 450.0, 0.0)


@pytest.fixture
def grid_def(lat_def, lon_def, hgt_def):
    return Grid(lat_def, lon_def, hgt_def)


@pytest.mark.parametrize('value, expected', [
    ('0', 0),
    ('1.0', 1),
    ('12.2', 12),
])
def test_coerce_to_int(value, expected):
    assert expected == IonexV1._coerce_into_int(value)


def test_coerce_to_int_error(ionex_file):
    inx = IonexV1(ionex_file)
    with pytest.raises(ValueError):
        inx._coerce_into_int('hello')


@pytest.mark.parametrize('epoch_str,dt', [
    ('''\
  1999     1     2     1     0     0                        EPOCH OF CURRENT MAP
''', datetime(1999, 1, 2, 1, 0, 0)),
    ('''\
  2000    12    12    11    20    15                        EPOCH OF CURRENT MAP
''', datetime(2000, 12, 12, 11, 20, 15)),
    ('''\
  2009     3     2    17    60     0                        EPOCH OF CURRENT MAP
''', datetime(2009, 3, 2, 18, 0, 0)),
    ('''\
  1996     3     9    20     0    60                        EPOCH OF CURRENT MAP
''', datetime(1996, 3, 9, 20, 1, 0)),
    ('''\
  1996     3     9    20    60    60                        EPOCH OF CURRENT MAP
''', datetime(1996, 3, 9, 21, 1, 0)),
    ('''\
  1996     3     9    20    61    73                        EPOCH OF CURRENT MAP
''', datetime(1996, 3, 9, 21, 2, 13)),
    # XXX: не все соблюдают формат
    ('''\
  2012     3     8     6     0  0.00                        EPOCH OF CURRENT MAP
''', datetime(2012, 3, 8, 6, 0, 0)),
    ('''\
  1997     4    16    24     0     0                        EPOCH OF CURRENT MAP
    ''', datetime(1997, 4, 17, 0, 0, 0)),
])
def test_parse_epoch(epoch_str, dt, ionex_file):
    inx = IonexV1(ionex_file)
    assert dt == inx._parse_epoch(epoch_str)


@pytest.mark.parametrize('grid_def_str,result', [
    ('''\
    87.5-180.0 180.0   5.0 450.0                            LAT/LON1/LON2/DLON/H
''', MapGridDef(87.5, -180., 180., 5., 450.)),
])
def test_parse_map_grid_def(grid_def_str, result):
    assert result == IonexV1._parse_map_grid_def(grid_def_str)


@pytest.mark.parametrize('line,result', [
    ('''\
   91   95   93   92   93   99  106  112  111
''', [91, 95, 93, 92, 93, 99, 106, 112, 111, ]),
    ('''\
   82   80   80   81   82   83   83   83   82   80   71   67   67   69   72   82
''', [82, 80, 80, 81, 82, 83, 83, 83, 82, 80, 71, 67, 67, 69, 72, 82, ]),
    ('''\
  864 1225 1850 2864 4459 6918106621632624853   82   85   89   94   96   94   91
''',
     [864, 1225, 1850, 2864, 4459, 6918, 10662, 16326, 24853, 82, 85,
      89, 94, 96, 94, 91]),
])
def test_read_slice(line, result):
    assert result == IonexV1._read_slice(line)


def test_read_header(ionex_header, lon_def, lat_def, hgt_def, grid_def):
    inx = IonexV1(ionex_header)
    with inx._context_manager as file_object:
        inx._read_header(file_object)

        # ожидаемые атрибуты
        assert inx.exponent == -1
        assert inx.dimension == 2

        assert inx.latitude == lat_def
        assert inx.longitude == lon_def
        assert inx.height == hgt_def

        assert inx.grid == grid_def

        # прервались сразу после заголовка
        line = next(file_object)
        assert line == (
            '     1                                                   '
            '   START OF TEC MAP\n'
        )


def test_read_map(one_map_file, one_map_file_data):
    inx = IonexV1(one_map_file)

    # FIXME: в фикстуры
    epoch = datetime(1999, 1, 2, 1, 0, 0)
    height = 450.0
    result = Map(epoch=epoch, height=height, data=one_map_file_data)

    # находимся в начале файла: промотаем до начала карты
    line = ''
    with inx._context_manager as file_object:
        for _ in range(19):
            line = next(file_object)
        assert line == '''\
     1                                                      START OF TEC MAP
'''
        assert result == inx._read_map(file_object)


def test_reader(ionex_file):
    inx = IonexV1(ionex_file)
    # имитируем чтение
    for _ in inx:
        pass
    assert inx._tec_maps_numbers == list(range(1, 13))


def test_reader_no_end_of_file(ionex_file_no_end_of_file):
    inx = IonexV1(ionex_file_no_end_of_file)
    with pytest.warns(UserWarning, match='Unexpected end of the file'):
        for _ in inx:
            pass
    assert inx._tec_maps_numbers == list(range(1, 13))


tests/test_reader.py
import pytest

from ionex import _get_version_type, reader, IonexV1


@pytest.mark.parametrize('line,expected_ver,expected_type', [
    ('''\
123456.1xx          fONOSPHERE MAPS     GPS                 IONEX VERSION / TYPE
''', 123456.1, 'f'),
    ('''\
     1.0            IONOSPHERE MAPS     GPS                 IONEX VERSION / TYPE
    ''', 1.0, 'I')
])
def test_get_version_type(line, expected_ver, expected_type):
    ver, t = _get_version_type(line)
    assert ver == expected_ver
    assert t == expected_type


def test_reader(ionex_file):
    inx = reader(ionex_file)
    assert isinstance(inx, IonexV1)

    for _ in inx:
        pass

    assert inx._tec_maps_numbers == list(range(1, 13))

    # оставляем открытым
    if not isinstance(ionex_file, str):
        assert not ionex_file.closed


ionex/__init__.py
from .ionex_file import IonexV1, NullContext
from .exceptions import IONEXError
from .exceptions import IONEXUnexpectedEnd

__all__ = ['reader']


def _get_version_type(line):
    return float(line[:8]), line[20]


def reader(file):
    """Возвращает читалку файла в формате IONEX.
    Читалка - итерируемый объект, на каждой итерации возвращает экземпляр
    ``ionex_map.IonexMap`` очередной карты, прочитанной из файла.

    :type file: str | file-object
    :param file: Путь к файлу IONEX или объект файла.

    :raises IONEXError:
        Если неизвестный тип или версия переданного файла.

    :raises IONEXUnexpectedEnd:
        Неполный файл.

    :raises IONEXMapError:
        Если возникли ошибки при обработке карты.
    """
    readers = {
        1.0: IonexV1,
    }

    if isinstance(file, str):
        context_manager = open(file)
    else:
        context_manager = NullContext(file)

    with context_manager as file_object:
        try:
            file_ver, file_type = _get_version_type(next(file_object))
        except StopIteration:
            raise IONEXUnexpectedEnd(file_object)

        if file_type != 'I':
            raise IONEXError('Unknown file type.')
        if file_ver not in readers:
            raise IONEXError('Unsupported version: {}'.format(file_ver))

        reader_class = readers[file_ver]
        return reader_class(file)


ionex/ionex_map.py
from collections import namedtuple

from .exceptions import IONEXMapError

Grid = namedtuple('Grid', ['latitude', 'longitude'])
Latitude = namedtuple('Latitude', ['lat1', 'lat2', 'dlat'])
Longitude = namedtuple('Longitude', ['lon1', 'lon2', 'dlon'])


class IonexMap:
    """Класс карты IONEX. Содержит карту и мета-данные.

    Атрибуты:

    :type grid: namedtuple
    :param grid: определение сетки для карты, содержит два
        ``namedtupla``:

        - ``grid.latittude`` =
          ``namedtuple('Latitude', ['lat1', 'lat2', 'dlat'])`` определение
          сетки по широте от ``lat1`` до ``lat2`` с шагом ``dlat``;

        - ``grid.longitude`` =
          ``namedtuple('Longitude', ['lon1', 'lon2', 'dlon'])``
          определение сетки по долготе от ``lon1`` до ``lon2`` с шагом
          ``dlon``.

    :type tec: list
    :param tec: данные ПЭС; одномерный список, представляет собой
        набор широтных "срезов" со значениями ПЭС.

        Начало каждого среза соответствует широте ``grid.latitude.lat1``, конец
        -- ``grid.latitude.lat2``, с шагом, равным
        ``grid.latitude.dlat``.

        Долгота первого среза соответствует ``grid.longitude.lon1``, долгота
        последнего -- ``grid.longitude.lon2``, с шагом, равным
        ``grid.longitude.dlon``.

    :type height: float
    :param height: высота, с которой ассоциированы данные карты.

    :type epoch: datetime
    :param epoch: дата и время карты ПЭС.
    """

    def __init__(self, *,
                 exponent,
                 epoch,
                 longitude,
                 latitude,
                 height,
                 tec,
                 rms=None,
                 none_value=None):
        """
        :param exponent:
            ``int``, значение 'EXPONENT' из файла IONEX; степень,
            в которую будут возведены значения ПЭС.

        :param epoch:
            ``datetime.datetime``, дата и время текущей карты.

        :param longitude:
            ``tuple``, определение сетки по широте, (lon1, lon2, dlon),
            предполагается, что значения совпадают со значениями заголовка
            'LON1 / LON2 / DLON' из файла IONEX.

        :param latitude:
           ``tuple``, определение сетки по широте, (lat1, lat2, dlat),
           предполагается, что значения совпадают со значениями заголовка
           'LAT1 / LAT2 / DLAT' из файла IONEX.

        :param height:
            ``float``, высота текущей карты.

        :param tec:
            ``list``, список значений ПЭС из файла IONEX.

        :param rms:
            ``list``, список значений RMS из файла IONEX.

        :param none_value:
            ``int``, значения в карте, равные ``none_value`` будут заменены на
            ``None``. По умолчанию ``none_value`` == None, в таком случае
            никакие замены производится не будут.
        """
        self.epoch = epoch
        self.height = height
        self.grid = Grid(
            latitude=Latitude(*latitude),
            longitude=Longitude(*longitude),
        )

        self._exponent = exponent
        self._none_value = none_value

        self._tec = tec.copy()
        self._rms = rms.copy() if rms is not None else None

        if not self._grid_match_data():
            err_msg = 'The grid definition does ' \
                      'not match the map; epoch {}.'.format(self.epoch)
            raise IONEXMapError(err_msg)

    @property
    def tec(self):
        """Вернуть ПЭС с учётом степени."""
        tec = [v * 10 ** self._exponent for v in self._tec]
        if self._none_value is None:
            return tec

        none_value = self._none_value * 10 ** self._exponent
        while True:
            try:
                i = tec.index(none_value)
                tec[i] = None
            except ValueError:
                break
        return tec

    @property
    def rms(self):
        raise NotImplementedError

    def _grid_match_data(self):
        def cells(start, stop, step):
            return (abs(start) + abs(stop)) / abs(step) + 1

        lat_cells = cells(*self.grid.latitude)
        lon_cells = cells(*self.grid.longitude)
        return lon_cells * lat_cells == len(self._tec)


ionex/ionex_file.py
import warnings
from collections import namedtuple
from datetime import datetime, timedelta

from .exceptions import IONEXUnexpectedEnd
from .ionex_map import IonexMap

Grid = namedtuple('Grid', ['latitude', 'longitude', 'height'])
Latitude = namedtuple('Latitude', ['lat1', 'lat2', 'dlat'])
Longitude = namedtuple('Longitude', ['lon1', 'lon2', 'dlon'])
Height = namedtuple('Height', ['hgt1', 'hgt2', 'dhgt'])

Map = namedtuple('Map', ['epoch', 'height', 'data'])
MapGridDef = namedtuple('MapGridDef', ['lat', 'lon1', 'lon2', 'dlon', 'h'])


class NullContext:
    def __init__(self, file):
        self.file = file

    def __enter__(self):
        return self.file

    def __exit__(self, exc_type, exc_val, exc_tb):
        pass


class IonexV1:
    # заголовки, которые нужно считать и соответствующие им атрибуты класса
    header_label = {
        'EXPONENT': 'exponent',
        'MAP DIMENSION': 'dimension',
        'LAT1 / LAT2 / DLAT': 'latitude',
        'LON1 / LON2 / DLON': 'longitude',
        'HGT1 / HGT2 / DHGT': 'height',
    }

    # "Non-available TEC values are written as '9999'" (описание IONEX)
    none_value = 9999

    def __init__(self, file):
        self._exponent = -1
        self._dimension = None

        self._lat = None
        self._lon = None
        self._height = None

        self._tec_maps_numbers = []

        if isinstance(file, str):
            self._context_manager = open(file)
        else:
            self._context_manager = NullContext(file)

    @property
    def exponent(self):
        return self._exponent

    @exponent.setter
    def exponent(self, value):
        self._exponent = int(value[0:6])

    @property
    def dimension(self):
        return self._dimension

    @dimension.setter
    def dimension(self, value):
        self._dimension = int(value[0:6])

    @property
    def latitude(self):
        return self._lat

    @latitude.setter
    def latitude(self, value):
        self._lat = Latitude(
            float(value[2:8]),
            float(value[8:14]),
            float(value[14:20]),
        )

    @property
    def longitude(self):
        return self._lon

    @longitude.setter
    def longitude(self, value):
        self._lon = Longitude(
            float(value[2:8]),
            float(value[8:14]),
            float(value[14:20]),
        )

    @property
    def height(self):
        return self._height

    @height.setter
    def height(self, value):
        self._height = Height(
            float(value[2:8]),
            float(value[8:14]),
            float(value[14:20]),
        )

    @property
    def grid(self):
        return Grid(self._lat, self._lon, self._height)

    @staticmethod
    def _get_label(line):
        return line[60:].rstrip()

    def _read_header(self, file_object):
        label = ''
        while label != 'END OF HEADER':
            try:
                line = next(file_object)
            except StopIteration:
                raise IONEXUnexpectedEnd(file_object)

            label = self._get_label(line)
            if label not in self.header_label:
                continue
            setattr(self, self.header_label[label], line)

    @staticmethod
    def _coerce_into_int(value):
        try:
            result = int(value)
        except ValueError:
            result = int(float(value))
            warnings.warn('Coerced into integer: {}'.format(value))
        return result

    def _parse_epoch(self, epoch_str):
        epoch_elements = []
        for i in range(0, 36, 6):
            # XXX: не все соблюдают формат, иногда попадаются float
            v = self._coerce_into_int(epoch_str[i:i + 6])
            epoch_elements.append(v)
        epoch = datetime(*epoch_elements[0:3])

        # иногда значение часа = 24 или минуты/секунды = 60
        # прибавляем как дельту
        epoch += timedelta(hours=epoch_elements[3])
        epoch += timedelta(seconds=epoch_elements[4] * 60 + epoch_elements[5])

        return epoch

    @staticmethod
    def _parse_map_grid_def(def_str):
        grid_def = []
        for i in range(2, 30, 6):
            v = float(def_str[i:i + 6])
            grid_def.append(v)
        return MapGridDef(*grid_def)

    @staticmethod
    def _read_slice(line):
        # 5 -- "ширина" поля со значением ПЭС,
        # поэтому длинна строки должна быть кратна 5
        line = line.rstrip()
        assert not len(line) % 5
        return [int(line[i:i+5]) for i in range(0, len(line), 5)]

    def _read_map(self, file_object):
        """
        :return: ``namedtuple``, Map('Map', ['epoch', 'height', 'data'])
        """
        epoch = 'EPOCH OF CURRENT MAP'
        grid = 'LAT/LON1/LON2/DLON/H'

        # FIXME: сразу после начала карты может быть EXPONENT
        #        объединить с header_label (?)
        parser = {
            epoch: self._parse_epoch,
            grid: self._parse_map_grid_def,
        }

        metadata = {
            epoch: None,
            grid: None,
        }

        data = []
        while True:
            try:
                line = next(file_object)
            except StopIteration:
                raise IONEXUnexpectedEnd(file_object)

            label = self._get_label(line)
            if label in parser:
                metadata[label] = parser[label](line)
                continue
            # TODO: проверять номер карты (?)
            elif label == 'END OF TEC MAP':
                break
            data += self._read_slice(line)

        return Map(
            epoch=metadata[epoch],
            # XXX: всегда будет заключительное значение высоты
            height=metadata[grid][4],
            data=data,
        )

    def _next_map(self):
        with self._context_manager as file_object:
            self._read_header(file_object)

            while True:
                try:
                    line = next(file_object).rstrip()
                except StopIteration:
                    filename = getattr(file_object, 'name', '<Unknown>')
                    warnings.warn(
                        'Unexpected end of the file {}.'.format(filename)
                    )
                    break

                label = self._get_label(line)
                if label == 'START OF TEC MAP':
                    self._tec_maps_numbers.append(int(line[:6]))
                    epoch, height, map_data = self._read_map(file_object)
                    yield IonexMap(
                        exponent=self.exponent,
                        epoch=epoch,
                        longitude=self.longitude,
                        latitude=self.latitude,
                        height=self.height,
                        tec=map_data,
                        none_value=self.none_value,
                    )
                    continue

                if label == 'END OF FILE':
                    break

    def __iter__(self):
        return self._next_map()
